components/Buttons/src/MDCFloatingButton.m

```diff
@@ -42,104 +55,442 @@ + (CGFloat)miniDimension {
   return MDCFloatingButtonMiniDimension;
 }
 
-+ (instancetype)floatingButtonWithShape:(MDCFloatingButtonShape)shape {
-  return [[[self class] alloc] initWithFrame:CGRectZero shape:shape];
-}
 
 - (instancetype)init {
-  return [self initWithFrame:CGRectZero shape:MDCFloatingButtonShapeDefault];
+  return [self initWithFrame:CGRectZero type:MDCFloatingButtonTypeDefault];
 }
 
 - (instancetype)initWithFrame:(CGRect)frame {
-  return [self initWithFrame:frame shape:MDCFloatingButtonShapeDefault];
+  return [self initWithFrame:frame type:MDCFloatingButtonTypeDefault];
 }
 
-- (instancetype)initWithFrame:(CGRect)frame shape:(MDCFloatingButtonShape)shape {
+- (instancetype)initWithFrame:(CGRect)frame type:(MDCFloatingButtonType)type {
   self = [super initWithFrame:frame];
   if (self) {
-    _shape = shape;
+    _type = type;
     // The superclass sets contentEdgeInsets from defaultContentEdgeInsets before the _shape is set.
     // Set contentEdgeInsets again to ensure the defaults are for the correct shape.
-    self.contentEdgeInsets = [self defaultContentEdgeInsets];
-    self.hitAreaInsets = [self defaultHitAreaInsets];
+    [self commonMDCFloatingButtonInit];
+    [self updateType];
   }
   return self;
 }
 
+- (void)commonMDCFloatingButtonInit {
+  const CGSize miniSizeNormal = CGSizeMake(40, 40);
+  const CGSize defaultSizeNormal = CGSizeMake(56, 56);
+  const CGSize defaultSizeExtendedMinimum = CGSizeMake(132, 48);
+  const CGSize defaultSizeExtendedMaximum = CGSizeMake(328, 0);
+
+
+  _typetoModeToMinimumSize = [NSMutableDictionary dictionaryWithCapacity:2];
+  _typetoModeToMinimumSize
+      = [@{
```

Comment:
I'd love for a comment since this is so unusual looking.

Reply:
Also, this looks like the kind of work that should have a dispatch once on it.

Reply:
Agreed. Once we settle on the baseline, it should be extracted out into a static bit of code. Instances can then use it to initialize the properties.